### PR TITLE
JBEHAVE-1489 Add ability to build converters chain by their source and target types

### PR DESCRIPTION
**Pull request by uarlouski**

### CODE REVIEW

**Review by ikalinin1 — APPROVED** [review]

File: jbehave-core/src/main/java/org/jbehave/core/steps/ParameterConverters.java [hunk 1]
```diff
@@ -572,29 +587,43 @@ public ParameterConversionFailed(String message, Throwable cause) {
...
-                return isAssignableFrom((Class<?>) targetType, type);
+            return areCompatible(targetType, type);
+        }
+
+        @Override
+        public boolean canConvertFrom(Type type) {
+            return areCompatible(sourceType, type);
+        }
+
+        private static boolean areCompatible(Type type1, Type type2) {
```
Comment (valfirst) [inline jbehave-core/src/main/java/org/jbehave/core/steps/ParameterConverters.java]: naming is not clear

File: jbehave-core/src/main/java/org/jbehave/core/steps/ParameterConverters.java [hunk 2]
```diff
@@ -572,29 +587,43 @@ public ParameterConversionFailed(String message, Throwable cause) {
         public AbstractParameterConverter() {
         }
 
-        public AbstractParameterConverter(Type targetType) {
-            super(targetType);
+        public AbstractParameterConverter(Type sourceType, Type targetType) {
```
Comment (valfirst) [inline jbehave-core/src/main/java/org/jbehave/core/steps/ParameterConverters.java]: isn't `sourceType` always equal to `String.class`  here?

File: jbehave-core/src/test/java/org/jbehave/core/steps/ParameterConvertersBehaviour.java [hunk 3]
```diff
@@ -1045,4 +1063,13 @@ public StringContainer convertValue(String value, Type type) {
...
         }
     }
+
+    private class ListToSetConverter extends AbstractChainableParameterConverter<List<String>, Set<String>> {
+
+        @Override
+        public Set<String> convertValue(List<String> value, Type type) {
+            return new HashSet<>(value);
+        }
+
```
Comment (valfirst) [inline jbehave-core/src/test/java/org/jbehave/core/steps/ParameterConvertersBehaviour.java:1072]: useless empty line

**Review by valfirst — CHANGES_REQUESTED** [review]

File: jbehave-core/src/test/java/org/jbehave/core/steps/ParameterConvertersBehaviour.java (hunk 3 above)
Comment (uarlouski) [reply jbehave-core/src/test/java/org/jbehave/core/steps/ParameterConvertersBehaviour.java:1072]: do we have convention regarding to this in jbehave? I see many classes with empty before the last curly bracket

File: jbehave-core/src/main/java/org/jbehave/core/steps/ParameterConverters.java (hunk 2 above)
Comment (uarlouski) [reply jbehave-core/src/main/java/org/jbehave/core/steps/ParameterConverters.java:590]: mm yes, that's why Functional parameter converter passes String.class here, I would prefer to keep this class decoupled from logic in Functional param converter

Comment (valfirst) [reply jbehave-core/src/main/java/org/jbehave/core/steps/ParameterConverters.java]: `... extends AbstractChainableParameterConverter<String, T> ...`

**Review by valfirst — APPROVED** [review]

**Comment by uarlouski** [comment]
In master https://github.com/jbehave/jbehave-core/commit/482b1ffd14f2187beddafed4503b3e6fb696b27e

**Pull request closed without merging**